Added basic intake subsystem and command

nrgill28 opened the pull request:
Intake is bound to LEFT BUMPER. Multiple conditions can be set so the intake only operates when they are all met. (e.g. The button is pressed and there is space for another ball)

Also the intake stuff won't do anything if the robot doesn't have an intake.

nrgill28 committed:
Added basic intake subsystem and command

Intake is bound to LEFT BUMPER. Multiple conditions can be set so the intake only operates when they are all met. (e.g. The button is pressed and there is space for another ball)

georgetzavelas commented on src/main/java/frc/robot/RobotContainer.java:
```diff
@@ -52,11 +55,13 @@ private void configureButtonBindings() {
-  
+    // Instantiate the intake command and bind it
+    intakeCommand = new IntakeCommand();
+    (new JoystickButton(driverStick, XboxController.Button.kBumperLeft.value)).whenHeld(intakeCommand);
```
Are the outer brackets needed here?

georgetzavelas commented on src/main/java/frc/robot/config/FluidConstant.java:
```diff
@@ -88,7 +97,7 @@ public void setValue(T newValue) {
     private void entryUpdated(EntryNotification notification) {
         // Only allow the value to be updated while the robot is disabled
-        if (DriverStation.getInstance().isDisabled()) {
+        if (!DriverStation.getInstance().isDisabled()) {
```
How about `DriverStation.getInstance().isEnabled()` to avoid a double negative?

georgetzavelas reviewed (commented):
I love the novel design using a HashMap! Some comments

nrgill28 committed:
2 Small fixes

nrgill28 commented:
Good catch! I didn't know there was a `isEnabled` as well as `isDisabled`. Guess I should have read the documentation 😜

kevlam2706 commented on src/main/java/frc/robot/config/Config.java:
```diff
@@ -68,6 +72,11 @@ private Config() {
+    // Fluid constants
+    static NetworkTable constantsTable = NetworkTableInstance.getDefault().getTable("constants");
+    public static FluidConstant<Double> INTAKE_SPEED = new FluidConstant<>("intake-target-speed", 0.25d)
+            .registerToTable(constantsTable);
```
Is it possible to define this constant inside the Intake subsystem and still have it be treated as a fluid constant?

I guess what I'm asking is if all constants for all subsystems should be defined in one common and global Config class or whether it makes more sense to define them inside the subsystems that use them.

Defining them in one spot makes them easy to find, but that breaks the encapsulation concept of having things contained in Subsystems.

nrgill28 replied:
Yes, it's possible to define them wherever you'd like. I originally put them in the config file because that is how it was done last year, although I would prefer having them in the subsystem classes.

kevlam2706 commented on src/main/java/frc/robot/commands/IntakeCommand.java:
```diff
@@ -0,0 +1,40 @@
+        addRequirements(intakeSubsystem);
+        
+        // Initialize the subsystem with this condition
+        intakeSubsystem.setCondition("button", false);
```
I wonder if there's a condition label that makes this a bit more clear, such as "activated" or "buttonPressed".

Could this same command theoretically also be used in autonomous or would there need to be a separate AutonomousIntake command?

kevlam2706 commented on src/main/java/frc/robot/subsystems/IntakeSubsystem.java:
```diff
@@ -0,0 +1,82 @@
+     */
+    public void setCondition(String name, boolean condition) {
+        conditions.put(name, condition);
+    }
```
This is cool. There is no need for it right now with the code you currently have, but this concept of conditions might be good for catching errors with the subsystems. For example, if there was an encoder in this subsystem, one of the run conditions would be that the encoder was initialized successfully. Same with any other mission critical hardware that cannot be missing or unreadable or Bad Things Happen.

kevlam2706 commented on src/main/java/frc/robot/subsystems/IntakeSubsystem.java:
```diff
@@ -0,0 +1,82 @@
+        // Check if all conditions are met.
+        // e.g. The intake button is held down AND there is space for more balls
+        boolean canRun = true;
+        for (var condition : conditions.values()) {
```
canRun = !conditions.containsValue(false)?

kevlam2706 reviewed (approved):
Looking good!

nrgill28 committed:
Conditions can now be ignored during auto/teleop

Also moved the FluidConstant for the intake speed to the intake subsystem and renamed the button condition to something more descriptive

nrgill28 committed:
Remaned IntakeCommand -> OperatorIntakeCommand

This is more descriptive

nrgill28 committed:
Renamed the ConditionActive enum to just Condition
